gtk4/src/expression.rs

```diff
@@ -0,0 +1,362 @@
+            unsafe fn set_value_optional(value: &mut glib::Value, this: Option<&Self>) {
+                ffi::gtk_value_set_expression(value.to_glib_none_mut().0, this.to_glib_none().0 as *mut _)
+            }
+        }
+
+    };
+}
+
+define_expression!(
+    PropertyExpression,
+    ffi::GtkPropertyExpression,
+    ffi::gtk_property_expression_get_type
+);
+
+impl PropertyExpression {
+    #[doc(alias = "gtk_property_expression_new")]
+    pub fn new<E: AsRef<Expression>>(
+        this_type: Type,
+        expression: Option<&E>,
+        property_name: &str,
+    ) -> Self {
+        skip_assert_initialized!();
+        unsafe {
+            from_glib_full(ffi::gtk_property_expression_new(
+                this_type.to_glib(),
+                expression.map(|e| e.as_ref()).to_glib_none().0,
+                property_name.to_glib_none().0,
+            ))
+        }
+    }
+
+    #[doc(alias = "gtk_property_expression_new_for_pspec")]
+    pub fn new_for_pspec<E: AsRef<Expression>>(
+        expression: Option<&E>,
+        pspec: glib::ParamSpec,
+    ) -> Self {
+        skip_assert_initialized!();
+        unsafe {
+            from_glib_full(ffi::gtk_property_expression_new_for_pspec(
+                expression.map(|e| e.as_ref()).to_glib_none().0,
+                pspec.to_glib_none().0,
+            ))
+        }
+    }
+
+    #[doc(alias = "gtk_property_expression_get_expression")]
+    pub fn get_expression(&self) -> Expression {
+        skip_assert_initialized!();
+        unsafe {
+            from_glib_none(ffi::gtk_property_expression_get_expression(
+                self.to_glib_none().0,
+            ))
+        }
+    }
+
+    #[doc(alias = "gtk_property_expression_get_pspec")]
+    pub fn get_pspec(&self) -> glib::ParamSpec {
+        skip_assert_initialized!();
+        unsafe {
+            from_glib_none(ffi::gtk_property_expression_get_pspec(
+                self.to_glib_none().0,
+            ))
+        }
+    }
+}
+
+define_expression!(
+    ObjectExpression,
+    ffi::GtkObjectExpression,
+    ffi::gtk_object_expression_get_type
+);
+
+impl ObjectExpression {
+    #[doc(alias = "gtk_property_expression_new")]
+    pub fn new<T: IsA<Object>>(object: &T) -> Self {
+        skip_assert_initialized!();
+        unsafe {
+            from_glib_full(ffi::gtk_object_expression_new(
+                object.as_ref().to_glib_none().0,
+            ))
+        }
+    }
+
+    #[doc(alias = "gtk_object_expression_get_object")]
+    pub fn get_object(&self) -> Option<Object> {
+        skip_assert_initialized!();
+        unsafe { from_glib_none(ffi::gtk_object_expression_get_object(self.to_glib_none().0)) }
+    }
+}
+
+define_expression!(
+    ClosureExpression,
+    ffi::GtkClosureExpression,
+    ffi::gtk_closure_expression_get_type
+);
+
+impl ClosureExpression {
+    #[doc(alias = "gtk_closure_expression_new")]
+    pub fn new(value_type: Type, closure: glib::Closure, params: &[Expression]) -> Self {
```

Comment:
If you feel like it you can also add a variant here that takes a Rust closure. That would also allow you to get rid of the `Send + Sync` requirement of `glib::Closure` as this is GTK and everything is on the main thread anyway.

Reply:
glib::Closure has a local variant, I used it here so that we have a simpler way to use closure expression.